Lots of conflict error logs when deleting history of a large process instance

### <!-- Component -->

<!-- Zeebe- -->

### <!-- Affected version -->

<!-- -8.9 -->

### <!-- Severity -->

<!-- Low- -->

### <!-- Likelihood -->

<!-- High_ -->

### Description

Lots of spam regarding conflicts are logged when deleting a large process instance. This observed using OpenSearch, but likely exists on Elasticsearch as well.


```
Caused by: org.opensearch.client.transport.httpclient5.ResponseException: method [POST
], host [http: //localhost:9200], URI [/camunda-audit-log-8.9.0_*/_delete_by_query?ignore_unavailable=true&allow_no_indices=true], status line [HTTP/1.1 409 Conflict]
    {
        "took": 2042,
        "timed_out": false,
        "total": 833030,
        "deleted": 0,
        "batches": 1,
        "version_conflicts": 1000,
        "noops": 0,
        "retries": {
            "bulk": 0,
            "search": 0
        },
        "throttled_millis": 0,
        "requests_per_second": -1.0,
        "throttled_until_millis": 0,
        "failures": [
            {
                "index": "camunda-audit-log-8.9.0_",
                "id": "1-7297712",
                "cause": {
                    "type": "version_conflict_engine_exception",
                    "reason": "[1-7297712]: version conflict, required seqNo [1672924], primary term [2]. but no document was found",
                    "index": "camunda-audit-log-8.9.0_",
                    "shard": "0",
                    "index_uuid": "zDynMy_7RVq9QMvKNCJPlQ"
                },
                "status": 409
            },
            {
                "index": "camunda-audit-log-8.9.0_",
                "id": "1-7298076",
                "cause": {
                    "type": "version_conflict_engine_exception",
                    "reason": "[1-7298076]: version conflict, required seqNo [1672925], primary term [2]. but no document was found",
                    "index": "camunda-audit-log-8.9.0_",
                    "shard": "0",
                    "index_uuid": "zDynMy_7RVq9QMvKNCJPlQ"
                },
                "status": 409
            },
....
```

### Steps to reproduce

1. Create a process instance with a lot of executed tasks (~1 million in my test)
2. Delete the history for this instance

### Current behavior

Lots of conflict error are logged. After some time the instance does get deleted properly.

### Expected behavior

The instance gets deleted without this additional noise.

### Workaround

1. do this
2. do that
3. now this happens and circumvents the bug


### Environment

SaaS & SM

### Version

- Component Version: `main` branch

### Rootcause

_No response_

### Solution ideas

We should be able to ignore conflicts and have the exporter retry the request

### Dev -> QA handover

- Resources:
- Versions to validate:
- Release version (in which version this feature will be released):


### Links

_No response_

### Automated Test Impact

_No response_

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/BackgroundTaskManagerFactory.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/BackgroundTaskManagerFactory.java
@@ -42,7 +42,6 @@ import io.camunda.search.connect.es.ElasticsearchConnector;
 import io.camunda.search.connect.os.OpensearchConnector;
 import io.camunda.webapps.schema.descriptors.DecisionInstanceDependant;
 import io.camunda.webapps.schema.descriptors.ProcessInstanceDependant;
-import io.camunda.webapps.schema.descriptors.index.HistoryDeletionIndex;
 import io.camunda.webapps.schema.descriptors.template.BatchOperationTemplate;
 import io.camunda.webapps.schema.descriptors.template.DecisionInstanceTemplate;
 import io.camunda.webapps.schema.descriptors.template.FlowNodeInstanceTemplate;
@@ -414,12 +413,7 @@ public final class BackgroundTaskManagerFactory {
 
     return buildHistoryDeletionTask(
         new HistoryDeletionJob(
-            dependantTemplates,
-            executor,
-            historyDeletionRepository,
-            logger,
-            resourceProvider.getIndexDescriptor(HistoryDeletionIndex.class),
-            resourceProvider.getIndexTemplateDescriptor(ListViewTemplate.class)));
+            dependantTemplates, executor, historyDeletionRepository, logger, resourceProvider));
   }
 
   private ReschedulingTask buildHistoryDeletionTask(final BackgroundTask task) {

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/ElasticsearchHistoryDeletionRepository.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/ElasticsearchHistoryDeletionRepository.java
@@ -8,6 +8,7 @@
 package io.camunda.exporter.tasks.historydeletion;
 
 import co.elastic.clients.elasticsearch.ElasticsearchAsyncClient;
+import co.elastic.clients.elasticsearch._types.Conflicts;
 import co.elastic.clients.elasticsearch._types.FieldValue;
 import co.elastic.clients.elasticsearch._types.SortOrder;
 import co.elastic.clients.elasticsearch.core.BulkRequest;
@@ -81,6 +82,8 @@ public class ElasticsearchHistoryDeletionRepository extends ElasticsearchReposit
             .index(sourceIndexName)
             .allowNoIndices(true)
             .ignoreUnavailable(true)
+            .waitForCompletion(true)
+            .conflicts(Conflicts.Proceed)
             .query(
                 q ->
                     q.terms(
@@ -103,6 +106,21 @@ public class ElasticsearchHistoryDeletionRepository extends ElasticsearchReposit
                 logger.error(errorMessage);
                 return CompletableFuture.failedFuture(new RuntimeException(errorMessage));
               }
+
+              if (response.versionConflicts() != null && response.versionConflicts() > 0) {
+                final var debugMessage =
+                    "Deleting documents from index '%s' by field '%s' encountered '%d' conflicts"
+                        .formatted(sourceIndexName, idFieldName, response.versionConflicts());
+                logger.debug(debugMessage);
+                return CompletableFuture.failedFuture(new RuntimeException(debugMessage));
+              }
+
+              logger.debug(
+                  "Deleted {} documents from index '{}' by field '{}' with values: {}",
+                  response.total(),
+                  sourceIndexName,
+                  idFieldName,
+                  fieldValues);
               return CompletableFuture.completedFuture(fieldValues);
             },
             executor);

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
@@ -16,6 +16,7 @@ import io.camunda.webapps.schema.descriptors.template.AuditLogTemplate;
 import io.camunda.webapps.schema.descriptors.template.ListViewTemplate;
 import io.camunda.webapps.schema.descriptors.template.OperationTemplate;
 import io.camunda.zeebe.protocol.record.value.HistoryDeletionType;
+import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
@@ -128,10 +129,10 @@ public class HistoryDeletionJob implements BackgroundTask {
 
   // TODO add javadoc
   private CompletionStage<List<String>> deleteProcessDefinitions(
-      final HistoryDeletionBatch batch, final List<String> ids) {
+      final HistoryDeletionBatch batch, final List<String> processInstanceDeletionIds) {
     final var processDefinitions = batch.getResourceKeys(HistoryDeletionType.PROCESS_DEFINITION);
     if (processDefinitions.isEmpty()) {
-      return CompletableFuture.completedFuture(ids);
+      return CompletableFuture.completedFuture(processInstanceDeletionIds);
     }
 
     return deleterRepository
@@ -144,7 +145,12 @@ public class HistoryDeletionJob implements BackgroundTask {
                 deleterRepository.deleteDocumentsById(
                     processIndex.getFullQualifiedName(),
                     processDefinitions.stream().map(Object::toString).toList()))
-        .thenApply(ignored -> batch.getHistoryDeletionIds(HistoryDeletionType.PROCESS_DEFINITION));
+        .thenApply(
+            ignored -> {
+              final var ids = new ArrayList<>(processInstanceDeletionIds);
+              ids.addAll(batch.getHistoryDeletionIds(HistoryDeletionType.PROCESS_DEFINITION));
+              return ids;
+            });
   }
 
   private CompletionStage<Integer> deleteFromHistoryDeletionIndex(final List<String> ids) {

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
@@ -7,6 +7,7 @@
  */
 package io.camunda.exporter.tasks.historydeletion;
 
+import io.camunda.exporter.ExporterResourceProvider;
 import io.camunda.exporter.tasks.BackgroundTask;
 import io.camunda.webapps.schema.descriptors.ProcessInstanceDependant;
 import io.camunda.webapps.schema.descriptors.index.HistoryDeletionIndex;
@@ -42,8 +43,7 @@ public class HistoryDeletionJob implements BackgroundTask {
       final Executor executor,
       final HistoryDeletionRepository historyDeletionRepository,
       final Logger logger,
-      final HistoryDeletionIndex historyDeletionIndex,
-      final ListViewTemplate listViewTemplate) {
+      final ExporterResourceProvider resourceProvider) {
     this.processInstanceDependants =
         processInstanceDependants.stream()
             .sorted(Comparator.comparing(ProcessInstanceDependant::getFullQualifiedName))
@@ -51,8 +51,8 @@ public class HistoryDeletionJob implements BackgroundTask {
     this.executor = executor;
     deleterRepository = historyDeletionRepository;
     this.logger = logger;
-    this.historyDeletionIndex = historyDeletionIndex;
-    this.listViewTemplate = listViewTemplate;
+    historyDeletionIndex = resourceProvider.getIndexDescriptor(HistoryDeletionIndex.class);
+    listViewTemplate = resourceProvider.getIndexTemplateDescriptor(ListViewTemplate.class);
   }
 
   @Override

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
@@ -11,6 +11,7 @@ import io.camunda.exporter.ExporterResourceProvider;
 import io.camunda.exporter.tasks.BackgroundTask;
 import io.camunda.webapps.schema.descriptors.ProcessInstanceDependant;
 import io.camunda.webapps.schema.descriptors.index.HistoryDeletionIndex;
+import io.camunda.webapps.schema.descriptors.index.ProcessIndex;
 import io.camunda.webapps.schema.descriptors.template.ListViewTemplate;
 import io.camunda.webapps.schema.descriptors.template.OperationTemplate;
 import io.camunda.zeebe.protocol.record.value.HistoryDeletionType;
@@ -37,6 +38,7 @@ public class HistoryDeletionJob implements BackgroundTask {
   private final Logger logger;
   private final HistoryDeletionIndex historyDeletionIndex;
   private final ListViewTemplate listViewTemplate;
+  private final ProcessIndex processIndex;
 
   public HistoryDeletionJob(
       final List<ProcessInstanceDependant> processInstanceDependants,
@@ -53,6 +55,7 @@ public class HistoryDeletionJob implements BackgroundTask {
     this.logger = logger;
     historyDeletionIndex = resourceProvider.getIndexDescriptor(HistoryDeletionIndex.class);
     listViewTemplate = resourceProvider.getIndexTemplateDescriptor(ListViewTemplate.class);
+    processIndex = resourceProvider.getIndexDescriptor(ProcessIndex.class);
   }
 
   @Override
@@ -75,7 +78,9 @@ public class HistoryDeletionJob implements BackgroundTask {
       return CompletableFuture.completedFuture(0);
     }
 
-    return deleteProcessInstances(batch).thenCompose(this::deleteFromHistoryDeletionIndex);
+    return deleteProcessInstances(batch)
+        .thenCompose(ids -> deleteProcessDefinitions(batch, ids))
+        .thenCompose(this::deleteFromHistoryDeletionIndex);
   }
 
   /**
@@ -116,6 +121,23 @@ public class HistoryDeletionJob implements BackgroundTask {
                   dependentSourceIdx, dependentIdFieldName, processInstanceKeys);
             })
         .thenApply(ignored -> batch.getHistoryDeletionIds(HistoryDeletionType.PROCESS_INSTANCE));
+  }
+
+  // TODO add javadoc
+  private CompletionStage<List<String>> deleteProcessDefinitions(
+      final HistoryDeletionBatch batch, final List<String> ids) {
+    final var processDefinitions = batch.getResourceKeys(HistoryDeletionType.PROCESS_DEFINITION);
+    if (processDefinitions.isEmpty()) {
+      return CompletableFuture.completedFuture(ids);
+    }
+
+    // TODO audit log??
+
+    return deleterRepository
+        .deleteDocumentsById(
+            processIndex.getFullQualifiedName(),
+            processDefinitions.stream().map(Object::toString).toList())
+        .thenApply(ignored -> batch.getHistoryDeletionIds(HistoryDeletionType.PROCESS_DEFINITION));
   }
 
   private CompletionStage<Integer> deleteFromHistoryDeletionIndex(final List<String> ids) {

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
@@ -12,6 +12,7 @@ import io.camunda.exporter.tasks.BackgroundTask;
 import io.camunda.webapps.schema.descriptors.ProcessInstanceDependant;
 import io.camunda.webapps.schema.descriptors.index.HistoryDeletionIndex;
 import io.camunda.webapps.schema.descriptors.index.ProcessIndex;
+import io.camunda.webapps.schema.descriptors.template.AuditLogTemplate;
 import io.camunda.webapps.schema.descriptors.template.ListViewTemplate;
 import io.camunda.webapps.schema.descriptors.template.OperationTemplate;
 import io.camunda.zeebe.protocol.record.value.HistoryDeletionType;
@@ -39,6 +40,7 @@ public class HistoryDeletionJob implements BackgroundTask {
   private final HistoryDeletionIndex historyDeletionIndex;
   private final ListViewTemplate listViewTemplate;
   private final ProcessIndex processIndex;
+  private final AuditLogTemplate auditLogTemplate;
 
   public HistoryDeletionJob(
       final List<ProcessInstanceDependant> processInstanceDependants,
@@ -56,6 +58,7 @@ public class HistoryDeletionJob implements BackgroundTask {
     historyDeletionIndex = resourceProvider.getIndexDescriptor(HistoryDeletionIndex.class);
     listViewTemplate = resourceProvider.getIndexTemplateDescriptor(ListViewTemplate.class);
     processIndex = resourceProvider.getIndexDescriptor(ProcessIndex.class);
+    auditLogTemplate = resourceProvider.getIndexTemplateDescriptor(AuditLogTemplate.class);
   }
 
   @Override
@@ -131,12 +134,16 @@ public class HistoryDeletionJob implements BackgroundTask {
       return CompletableFuture.completedFuture(ids);
     }
 
-    // TODO audit log??
-
     return deleterRepository
-        .deleteDocumentsById(
-            processIndex.getFullQualifiedName(),
-            processDefinitions.stream().map(Object::toString).toList())
+        .deleteDocumentsByField(
+            auditLogTemplate.getFullQualifiedName() + "*",
+            AuditLogTemplate.PROCESS_DEFINITION_KEY,
+            processDefinitions)
+        .thenCompose(
+            ignored ->
+                deleterRepository.deleteDocumentsById(
+                    processIndex.getFullQualifiedName(),
+                    processDefinitions.stream().map(Object::toString).toList()))
         .thenApply(ignored -> batch.getHistoryDeletionIds(HistoryDeletionType.PROCESS_DEFINITION));
   }
 

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
@@ -137,7 +137,7 @@ public class HistoryDeletionJob implements BackgroundTask {
 
     return deleterRepository
         .deleteDocumentsByField(
-            auditLogTemplate.getFullQualifiedName() + "*",
+            auditLogTemplate.getIndexPattern(),
             AuditLogTemplate.PROCESS_DEFINITION_KEY,
             processDefinitions)
         .thenCompose(

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java
@@ -127,7 +127,17 @@ public class HistoryDeletionJob implements BackgroundTask {
         .thenApply(ignored -> batch.getHistoryDeletionIds(HistoryDeletionType.PROCESS_INSTANCE));
   }
 
-  // TODO add javadoc
+  /**
+   * This method will delete a batch of process definitions in two steps:
+   *
+   * <ol>
+   *   <li>It deletes the process definition related data from the audit logs
+   *   <li>It deletes the process definitions from the process index
+   * </ol>
+   *
+   * @param batch The batch of entities to delete
+   * @return A future containing the list of history-deletion IDs that were processed
+   */
   private CompletionStage<List<String>> deleteProcessDefinitions(
       final HistoryDeletionBatch batch, final List<String> processInstanceDeletionIds) {
     final var processDefinitions = batch.getResourceKeys(HistoryDeletionType.PROCESS_DEFINITION);

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/OpenSearchHistoryDeletionRepository.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/OpenSearchHistoryDeletionRepository.java
@@ -20,6 +20,7 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
 import javax.annotation.WillCloseWhenClosed;
 import org.opensearch.client.opensearch.OpenSearchAsyncClient;
+import org.opensearch.client.opensearch._types.Conflicts;
 import org.opensearch.client.opensearch._types.FieldValue;
 import org.opensearch.client.opensearch._types.SortOrder;
 import org.opensearch.client.opensearch.core.BulkRequest;
@@ -82,6 +83,8 @@ public class OpenSearchHistoryDeletionRepository extends OpensearchRepository
             .index(sourceIndexName)
             .allowNoIndices(true)
             .ignoreUnavailable(true)
+            .waitForCompletion(true)
+            .conflicts(Conflicts.Proceed)
             .query(
                 q ->
                     q.terms(
@@ -106,6 +109,22 @@ public class OpenSearchHistoryDeletionRepository extends OpensearchRepository
                         logger.error(errorMessage);
                         return CompletableFuture.failedFuture(new RuntimeException(errorMessage));
                       }
+
+                      if (response.versionConflicts() != null && response.versionConflicts() > 0) {
+                        final var debugMessage =
+                            "Deleting documents from index '%s' by field '%s' encountered '%d' conflicts"
+                                .formatted(
+                                    sourceIndexName, idFieldName, response.versionConflicts());
+                        logger.debug(debugMessage);
+                        return CompletableFuture.failedFuture(new RuntimeException(debugMessage));
+                      }
+
+                      logger.debug(
+                          "Deleted {} documents from index '{}' by field '{}' with values: {}",
+                          response.total(),
+                          sourceIndexName,
+                          idFieldName,
+                          fieldValues);
                       return CompletableFuture.completedFuture(fieldValues);
                     },
                     executor));

--- a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJobTest.java
+++ b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJobTest.java
@@ -11,6 +11,8 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyList;
 import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.atMostOnce;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.verify;
@@ -19,6 +21,8 @@ import static org.mockito.Mockito.when;
 import io.camunda.exporter.tasks.utils.TestExporterResourceProvider;
 import io.camunda.webapps.schema.descriptors.ProcessInstanceDependant;
 import io.camunda.webapps.schema.descriptors.index.HistoryDeletionIndex;
+import io.camunda.webapps.schema.descriptors.index.ProcessIndex;
+import io.camunda.webapps.schema.descriptors.template.AuditLogTemplate;
 import io.camunda.webapps.schema.descriptors.template.ListViewTemplate;
 import io.camunda.webapps.schema.descriptors.template.OperationTemplate;
 import io.camunda.webapps.schema.entities.HistoryDeletionEntity;
@@ -40,6 +44,8 @@ final class HistoryDeletionJobTest {
   private List<ProcessInstanceDependant> dependants;
   private HistoryDeletionIndex historyDeletionIndex;
   private ListViewTemplate listViewTemplate;
+  private ProcessIndex processIndex;
+  private AuditLogTemplate auditLogTemplate;
 
   @BeforeEach
   void setUp() {
@@ -51,6 +57,10 @@ final class HistoryDeletionJobTest {
             .filter(ProcessInstanceDependant.class::isInstance)
             .map(ProcessInstanceDependant.class::cast)
             .toList();
+    historyDeletionIndex = resourceProvider.getIndexDescriptor(HistoryDeletionIndex.class);
+    listViewTemplate = resourceProvider.getIndexTemplateDescriptor(ListViewTemplate.class);
+    processIndex = resourceProvider.getIndexDescriptor(ProcessIndex.class);
+    auditLogTemplate = resourceProvider.getIndexTemplateDescriptor(AuditLogTemplate.class);
     job = new HistoryDeletionJob(dependants, executor, repository, LOGGER, resourceProvider);
   }
 
@@ -191,6 +201,133 @@ final class HistoryDeletionJobTest {
             listViewTemplate.getIndexPattern(),
             ListViewTemplate.PROCESS_INSTANCE_KEY,
             List.of(entity.getResourceKey()));
+    verify(repository, never()).deleteDocumentsById(any(), any());
+  }
+
+  @Test
+  void shouldDeleteProcessDefinitionHistory() {
+    // given
+    final var entity =
+        new HistoryDeletionEntity()
+            .setId("id")
+            .setResourceKey(1L)
+            .setResourceType(HistoryDeletionType.PROCESS_DEFINITION)
+            .setBatchOperationKey(2L)
+            .setPartitionId(1);
+    when(repository.getNextBatch())
+        .thenReturn(CompletableFuture.completedFuture(new HistoryDeletionBatch(List.of(entity))));
+    when(repository.deleteDocumentsByField(anyString(), anyString(), anyList()))
+        .thenReturn(CompletableFuture.completedFuture(List.of()));
+    when(repository.deleteDocumentsById(anyString(), anyList()))
+        .thenReturn(CompletableFuture.completedFuture(0));
+
+    // when
+    job.execute().toCompletableFuture().join();
+
+    // then Process Definition is deleted
+    verify(repository, atMostOnce())
+        .deleteDocumentsByField(
+            auditLogTemplate.getIndexPattern(),
+            AuditLogTemplate.PROCESS_DEFINITION_KEY,
+            List.of(entity.getResourceKey()));
+    verify(repository, atMostOnce())
+        .deleteDocumentsById(
+            processIndex.getFullQualifiedName(), List.of(String.valueOf(entity.getResourceKey())));
+    verify(repository, atMostOnce())
+        .deleteDocumentsById(historyDeletionIndex.getFullQualifiedName(), List.of(entity.getId()));
+  }
+
+  @Test
+  void shouldNotDeleteProcessDefinitionIfAuditLogDeletionFailed() {
+    // given
+    final var entity =
+        new HistoryDeletionEntity()
+            .setId("id")
+            .setResourceKey(1L)
+            .setResourceType(HistoryDeletionType.PROCESS_DEFINITION)
+            .setBatchOperationKey(2L)
+            .setPartitionId(1);
+    when(repository.getNextBatch())
+        .thenReturn(CompletableFuture.completedFuture(new HistoryDeletionBatch(List.of(entity))));
+    when(repository.deleteDocumentsByField(anyString(), anyString(), anyList()))
+        .thenReturn(CompletableFuture.failedFuture(new RuntimeException("Failed deleting")));
+
+    // when
+    job.execute().exceptionally(ex -> 0).toCompletableFuture().join();
+
+    // then
+    verify(repository, never())
+        .deleteDocumentsById(
+            processIndex.getFullQualifiedName(), List.of(String.valueOf(entity.getResourceKey())));
+    verify(repository, never()).deleteDocumentsById(any(), any());
+  }
+
+  @Test
+  void shouldNotDeleteProcessDefinitionFromHistoryIndexIfDeletionFailed() {
+    // given
+    final var entity =
+        new HistoryDeletionEntity()
+            .setId("id")
+            .setResourceKey(1L)
+            .setResourceType(HistoryDeletionType.PROCESS_DEFINITION)
+            .setBatchOperationKey(2L)
+            .setPartitionId(1);
+    when(repository.getNextBatch())
+        .thenReturn(CompletableFuture.completedFuture(new HistoryDeletionBatch(List.of(entity))));
+    when(repository.deleteDocumentsByField(anyString(), anyString(), anyList()))
+        .thenReturn(CompletableFuture.completedFuture(List.of()));
+    verify(repository, atMostOnce())
+        .deleteDocumentsById(
+            processIndex.getFullQualifiedName(), List.of(String.valueOf(entity.getResourceKey())));
+
+    // when
+    job.execute().exceptionally(ex -> 0).toCompletableFuture().join();
+
+    // then
+    verify(repository, atMostOnce())
+        .deleteDocumentsByField(
+            auditLogTemplate.getIndexPattern(),
+            AuditLogTemplate.PROCESS_DEFINITION_KEY,
+            List.of(entity.getResourceKey()));
+    verify(repository, atMostOnce())
+        .deleteDocumentsById(
+            processIndex.getFullQualifiedName(), List.of(String.valueOf(entity.getResourceKey())));
+    verify(repository, never())
+        .deleteDocumentsById(eq(historyDeletionIndex.getFullQualifiedName()), any());
+  }
+
+  @Test
+  void shouldNotDeleteProcessDefinitionIfProcessInstanceDeletionFailed() {
+    // given
+    final var entity1 =
+        new HistoryDeletionEntity()
+            .setId("id1")
+            .setResourceKey(1L)
+            .setResourceType(HistoryDeletionType.PROCESS_INSTANCE)
+            .setBatchOperationKey(2L)
+            .setPartitionId(1);
+    final var entity2 =
+        new HistoryDeletionEntity()
+            .setId("id2")
+            .setResourceKey(2L)
+            .setResourceType(HistoryDeletionType.PROCESS_DEFINITION)
+            .setBatchOperationKey(2L)
+            .setPartitionId(1);
+    when(repository.getNextBatch())
+        .thenReturn(
+            CompletableFuture.completedFuture(new HistoryDeletionBatch(List.of(entity1, entity2))));
+    when(repository.deleteDocumentsByField(anyString(), anyString(), anyList()))
+        .thenReturn(CompletableFuture.completedFuture(List.of()));
+    when(repository.deleteDocumentsByField(
+            listViewTemplate.getIndexPattern(),
+            ListViewTemplate.PROCESS_INSTANCE_KEY,
+            List.of(entity1.getResourceKey())))
+        .thenReturn(CompletableFuture.failedFuture(new RuntimeException("Failed deleting")));
+
+    // when
+    job.execute().exceptionally(ex -> 0).toCompletableFuture().join();
+
+    // then
     verify(repository, never()).deleteDocumentsById(any(), any());
   }
 }

--- a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJobTest.java
+++ b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJobTest.java
@@ -276,9 +276,8 @@ final class HistoryDeletionJobTest {
         .thenReturn(CompletableFuture.completedFuture(new HistoryDeletionBatch(List.of(entity))));
     when(repository.deleteDocumentsByField(anyString(), anyString(), anyList()))
         .thenReturn(CompletableFuture.completedFuture(List.of()));
-    verify(repository, atMostOnce())
-        .deleteDocumentsById(
-            processIndex.getFullQualifiedName(), List.of(String.valueOf(entity.getResourceKey())));
+    when(repository.deleteDocumentsById(eq(processIndex.getFullQualifiedName()), any()))
+        .thenReturn(CompletableFuture.failedFuture(new RuntimeException("Failed deleting")));
 
     // when
     job.execute().exceptionally(ex -> 0).toCompletableFuture().join();

--- a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJobTest.java
+++ b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJobTest.java
@@ -46,16 +46,12 @@ final class HistoryDeletionJobTest {
     repository = mock(HistoryDeletionRepository.class);
     final TestExporterResourceProvider resourceProvider =
         new TestExporterResourceProvider("", true);
-    historyDeletionIndex = resourceProvider.getIndexDescriptor(HistoryDeletionIndex.class);
-    listViewTemplate = resourceProvider.getIndexTemplateDescriptor(ListViewTemplate.class);
     dependants =
         resourceProvider.getIndexTemplateDescriptors().stream()
             .filter(ProcessInstanceDependant.class::isInstance)
             .map(ProcessInstanceDependant.class::cast)
             .toList();
-    job =
-        new HistoryDeletionJob(
-            dependants, executor, repository, LOGGER, historyDeletionIndex, listViewTemplate);
+    job = new HistoryDeletionJob(dependants, executor, repository, LOGGER, resourceProvider);
   }
 
   @Test

